src/main/scala/is/hail/methods/IBD.scala

```diff
@@ -172,32 +167,75 @@ object IBD {
         gts.grouped(chunkSize)
           .zipWithIndex
           .map { case (gtGroup, i) => ((i, variantId / chunkSize), (vid, gtGroup)) }
-      }
+    }
       .aggregateByKey(Array.tabulate(chunkSize * chunkSize)((i) => IBSFFI.missingGTCRep))({ case (x, (vid, gs)) =>
         for (i <- gs.indices) x(vid * chunkSize + i) = gs(i)
         x
       }, { case (x, y) =>
-        for (i <- y.indices)
-          if (x(i) == IBSFFI.missingGTCRep)
-            x(i) = y(i)
-        x
+          for (i <- y.indices)
+            if (x(i) == IBSFFI.missingGTCRep)
+              x(i) = y(i)
+          x
       })
       .map { case ((s, v), gs) => (v, (s, IBSFFI.pack(chunkSize, chunkSize, gs))) }
 
     chunkedGenotypeMatrix.join(chunkedGenotypeMatrix)
-      // optimization: Ignore chunks below the diagonal
+    // optimization: Ignore chunks below the diagonal
       .filter { case (_, ((i, _), (j, _))) => j >= i }
       .map { case (_, ((s1, gs1), (s2, gs2))) =>
         ((s1, s2), IBSFFI.ibs(chunkSize, chunkSize, gs1, gs2))
-      }
+    }
       .reduceByKey { (a, b) =>
-        var i = 0
-        while (i != a.length) {
-          a(i) += b(i)
-          i += 1
-        }
-        a
+      var i = 0
+      while (i != a.length) {
+        a(i) += b(i)
+        i += 1
       }
+      a
+    }
+  }
+
+  private def selectFromFusedIbs(n: Int)(rowLimit: Int, colLimit: Int, fused: Array[Long]): Array[Double] = {
+    assert(n >= 0)
+    assert(n <= 2)
+    val arr = new Array[Double](rowLimit * colLimit)
+    var si = 0
+    var sj = 0
+    while (si < rowLimit) {
+      while (sj < colLimit) {
+        arr(sj*rowLimit + si) = fused(si * chunkSize * 3 + sj * 3 + n).toDouble
+        sj += 1
+      }
+      sj = 0
+      si += 1
+    }
+    arr
+  }
+  def ibs(vds: VariantDataset): (BlockMatrix, BlockMatrix, BlockMatrix) = {
+    val lastBlock = (vds.nSamples - 1) / chunkSize
+    val extras = vds.nSamples % chunkSize
+    def getIbsAsBlockMatrix(n: Int)(fused: RDD[((Int, Int), Array[Long])]): BlockMatrix =
+      new BlockMatrix(fused.map { case ((i, j), x) =>
+        val rows = if (i == lastBlock) extras else chunkSize
+        val cols = if (j == lastBlock) extras else chunkSize
+        val ibsn = selectFromFusedIbs(n)(rows, cols, x)
+        ((i, j), new DenseMatrix(rows, cols, ibsn, false): Matrix)
+      }, chunkSize, chunkSize)
+
+    val fused = fusedIbs(vds).cache()
```

Comment:
No reason, I'm not really sure which is better for this situation. I want to avoid re-running it three times for each IBS state, if all three are used. If we generally prefer persist, I can put a persist in.